File: JavaScript/04_Array-IV/M003. Longest Substring Without Repeating Characters.js
// 3. Longest Substring Without Repeating Characters
// https://leetcode.com/problems/longest-substring-without-repeating-characters/

/**
 * @param {string} s
 * @return {number}
 */
function lengthOfLongestSubstring(s) {
    const n = s.length;
    let maxLength = 0;
    let left = 0;
    const charIndexMap = new Map();

    for (let right = 0; right < n; right++) {
        const currentChar = s[right];
        if (charIndexMap.has(currentChar)) {
            left = Math.max(left, charIndexMap.get(currentChar) + 1);
        }
        charIndexMap.set(currentChar, right);
        maxLength = Math.max(maxLength, right - left + 1);
    }

    return maxLength;
}

const input = "abcabcbb";
console.log("Length of longest substring without repeating characters:", lengthOfLongestSubstring(input)); // Output: 3

File: Java/04_Array-IV/M003. Longest Substring Without Repeating Characters.java
// Longest Substring Without Repeating Characters.
// https://leetcode.com/problems/longest-substring-without-repeating-characters/

import java.util.*;

class Solution {

    public int lengthOfLongestSubstring(String s) {
        int n = s.length();
        int maxLength = 0;
        int left = 0;
        Map<Character, Integer> charIndexMap = new HashMap<>();

        for (int right = 0; right < n; right++) {
            char currentChar = s.charAt(right);
            if (charIndexMap.containsKey(currentChar)) {
                left = Math.max(left, charIndexMap.get(currentChar) + 1);
            }
            charIndexMap.put(currentChar, right);
            maxLength = Math.max(maxLength, right - left + 1);
        }

        return maxLength;
    }

    public static void main(String[] args) {
        Solution solution = new Solution();
        String input = "abcabcbb";
        int result = solution.lengthOfLongestSubstring(input);
        System.out.println(
            "Length of longest substring without repeating characters: " +
            result
        );
    }
}
